LSPS0/README.md

```diff
@@ -512,26 +514,127 @@ the LSP supports:
+> and reduce error sources. Additionally, the client will only need to call one method 
+> to negotiate all the LSPS versions, instead of one method per LSPS spec.
+
+
+LSPS0 MUST provide a method called `lsps0.select_versions`.
+This method takes the following parameters:
+
+```JSON
+{
+  "client_versions": {
+    "lsps0": [1],
+    "lsps1": [2, 3],
+    "lsps2": [1, 2, 3],
+    ...
+  }
+}
+```
+
+`client_versions` is a dictionary in the form of `"lsps_name": supported_client_versions` .
+- The client MUST list all `lsps_name` specs it will use for this connection session.
+- `supported_client_versions` is an array of integral non-zero positive JSON numbers, 
+in any order and represents a set of API versions that the client supports.
+
+The LSP MUST select the highest-numbered version in the
+`client_versions` that the LSP also supports.
+
+If the LSP is able to find a version in the `supported_client_versions`
+array that is supported by the LSP, then the LSP selects the
+highest-numbered version it can find as the "current client
+version", which it retains as long as the client is connected or until the client changes the versions.
+If the LSP does not support any of the versions, the LSP MUST return the number `-1`.
+If the client disconnects, the LSP forgets the current client
+version for that client.
+On connection, the LSP sets the current client version to `-1`,
+which is an invalid version, indicating that the
+spec API cannot be accessed.
+
+On success, the result of the `lsps0.select_versions` call
+is in the form of:
+
+```JSON
+{
+  "api_versions": {
+    "lsps0": 1,
+    "lsps1": 3,
+    "lsps2": -1,
+  }
+}
+```
+
+`api_versions` is a dictionary in the form of `"lsps_name": selected_api_version`.
+- It MUST contain all `lsps_name` specs that the client requested.
+- `selected_api_version` is the version the LSP selected among 
+the `supported_client_versions`.
+
+On connection establishment, the client MUST call `lsps0.select_versions`
+to negotiate the API versions with the LSP.
+
+> **Rationale** Some LSPS specifications may send 
+> notifications to the client. It is important that the LSP
+> knows which version of the spec the client is using, so it 
+> can send the correct notifications.
+
+The client MAY call `lsps0.list_protocols` 
+before calling `lsps0.select_versions` to determine which LSPS specs the LSP supports. This is not required though.
+
+Except for `lsps0.select_versions` and `lsps0.list_protocols`, any other LSPS specification method 
+call that does not have a negotiated version number, MUST fail with
+
+* `unsupported_version` (1) - The client and the LSP have not yet
+  negotiated a API version via `lsps0.select_versions`.
+
+`lsps0.select_versions` and `lsps0.list_protocols` are exempt from this error
+as they are needed to negotiate the API versions.
+
+`lsps0.select_versions` MAY be called multiple times during the same connection session.
```

Comment:
> As I have not heard a convincing argument yet why we should entirely overhaul the basic assumptions of the protocol (namely one node, one LSP, communicating via JSON-RPC over a single BOLT8/TCP connection), it's a NACK so far from me.

This is not done to support multiple apps. The assumption is still one node communicating with the LSPSpec and one version to be active at a time. The only thing I changed here is that a version upgrade/downgrade can happen without disconnecting/connecting the tcp connection as this could have detrimental effects on other services outside of the LSPSpec.

> namely one node, one LSP, communicating via JSON-RPC over a single BOLT8/TCP connection

Happy to write this base assumption in this PR if you wish. This PR is exclusively reserved for one node. If we were to support multiple apps/signer at a time, @ErikDeSmedt or I would create a new PR. I am not convinced yet this is needed tho and there are still discussions going on. Highly likely we will just stick with one node/app.

Reply:
I agree, we need to talk about this. That is why I added the migration section to this PR. I quote:

>Every LSPS SHOULD address version switching/migrations behaviour 
> if a version number is increased or decreased.

> > **Rationale** In case the client switches to a different version,
> > existing states might need to be abandoned or migrated.
> > It is important to clearly define these behaviours to avoid
> > confusion.

The thing is: This migration behaviour would have occurred anyway as soon as we have multiple versions of an LSPS. It doesn't matter if we now switch versions during the tpc-session or when disconnecting/reconnecting.

I personally think this should be addressed spec to spec because different specs have different requirements. Happy to discuss on the next call.

> Btw., on this note we should probably generally start talking about protocol-level compatibility guarantees, i.e., if we think some protocols should be resume operation after reconnect (and what implication that has on protocol-level changes) or if we generally require/assume every 'session' to be entirely independent, so we never have to care about upgrade paths between versions.

So far I assumed the protocol resumes operation after each session/reconnect. I don't see a reason why to do short lived session.